### DISCUSSION TITLE
Help with design a structure

### DISCUSSION THREAD

**Discussion by kmosakowski** [remkop/picocli#2166] [Q&A] [answered]
Hello all :)

I'm new in the Picocli world. I'm working on rewriting a JCommander framework to Picocli.
Why that decision? Because, in my opinion Picocli is more powerful than JCommander. But JCommander and Picocli it is something new for me.

I would like to ask you for help. I need to create a structure something like that:
![image](https://github.com/remkop/picocli/assets/32022222/07401522-e309-4096-8bd6-924b5dd7a251)

Right now, it looks like that:

AllArg.class:

- @ArgGroup Debug.class
- @ArgGroup Command1.class
- @ArgGroup Command2.class
- @ArgGroup Command3.class
- - @Options field1
- - @Options field2
- - @Options field3
- - @ArgGroup SubCommand1.class
- - @ArgGroup SubCommand2.class
- - @ArgGroup SubCommand3.class
- - @ArgGroup SubCommand4.class
- - @ArgGroup SubCommand5.class
- - @ArgGroup SubCommand6.class
- - @ArgGroup SubCommand7.class

What problem do I have?

1. Command1, Command2 and Command3 are not mutually exclusive. I don't know how to resolve it.
2. Important for me is auto generated help. That I use @ArgGroup. All SubCommand are first in my help, I would like to change its place and have it in the middle of the help. `sortOption = false` doesn't work
3. How to execute `public run()` in Command2.class through AllArg.class?
4. What is the best way to run the program and work on all commands? A huge wall of ifs and taking all fields and checking if they have value? What is your way?

I like very simple and 'clean code' solutions. I feel that I can't achieve this by myself, that I'm asking.

Thank you for your time, and have a nice day :)

**Comment by remkop** [comment, marked as answer]
Instead of `@ArgGroup`, I believe you want to use `@Command` instead.

From your diagram it sounds like you want something like this:

```java
@Command(subcommands = { Command1.class, Command2.class, Command3.class})
class Input { // maybe implements Runnable if this command also has functionality
    @Option(names = {"-d", "--debug"}, scope = ScopeType.INHERIT) // option is shared with subcommands
    boolean debug;
}

@Command
class Command1 implements Runnable { ... }

@Command
class Command2 implements Runnable { ... }

@Command(subcommands = { SubCommand1.class, SubCommand2.class, SubCommand3.class})
class Command3 implements Runnable { 
     @Option(names = "-x")
     String x;

     @Option(names = "-y")
     String y;

    public void run() {
        // business logic for Command 3 here
    }
}
```

**Reply by kmosakowski** [reply]
I would like to try your way. But the first problem what I see is Usage help. I use `resourceBudle` to do some header. and my output looks like that:

```
MY CUSTOM HEADER
...
MY CUSTOM HEADER

Commands:
  Command1  ____  _     _____ _ _ _
```
  
Without it, looks like that

```
MY CUSTOM HEADER
...
MY CUSTOM HEADER

Commands:
  Command1  Description for Command1
```

The Main Argument class:

```
@Command(requiredOptionMarker = '*',
        sortOptions = false,
        resourceBundle = "help",
        usageHelpAutoWidth = true,
        subcommands = {Command1.class}
)
public class ApplicationArgument {

    @Option(names = {"-d", "--debug"}, scope = ScopeType.INHERIT) // option is shared with subcommands
    boolean debug;
}
```

**Reply by remkop** [reply]
What does your resourcebundle look like? 
Java 9 supports UTF-8 resource bundles, but if you are running on Java 8 or older, then you'd need to use the [native2ascii](http://docs.oracle.com/javase/7/docs/technotes/tools/windows/native2ascii.html) tool to convert an UTF-8 saved properties file to an ISO-8859-1 saved properties file wherein all uncovered characters are converted into `\uXXXX` format.

**Reply by kmosakowski** [reply]
I use java 21. My resource bundles looks like that:

```
usage.header   =____  _ ____|||
usage.header.0 =| | ___}}||
usage.header.1 =| | LOGO
usage.header.2 =| | ____||||
usage.header.3 =|_|_|||||
usage.header.4 =
usage.header.5 =
usage.header.6 =Some text
usage.header.7 =
usage.header.8 =--------separation
usage.header.9 =Some text
usage.header.10=
usage.synopsisHeading = Usage:%n
usage.customSynopsis.0 =\u0020 some text
usage.customSynopsis.1 =\u0020 some text
usage.customSynopsis.2 =\u0020 some text
usage.customSynopsis.3 =\u0020 some text
```

The problem is only with Commands. The options, header and so on, looks well

**Reply by kmosakowski** [reply]
I think, I understand what is going on with help and resource bundles. Please look at it:

help.properties:
```
usage.header.0=test
usage.header.1=test2
usage.header.2=test3
```

Usage in Console:
```
test
test2
test3
Usage: APP [-hV] [COMMAND]
  -h, --help      Show this help message and exit.
  -V, --version   Print version information and exit.
Commands:
  run     test
  import  test
 
...
```

Description of every COMMAND is taken from help.properties from first line. Check where the word 'test' is repeated.

**Reply by remkop** [reply]
The subcommands list show only the first line of the usage header, this is by design.
You can give subommands different headers by prefixing their header with the command name:

```
#resourcebundle
usage.header.0=test
usage.header.1=test2
usage.header.2=test3
run.usage.header.0=Run test
import.usage.header.0=Import test
```
